Sopel 8 upgrade guide

dgw opened the pull request:
It's time that this branch had a PR, so other maintainers can give feedback on the upgrade guide ahead of release. Draft for now; goodness knows the history already needs cleanup, and it will need even more after review(s).

dgw committed:
upgrading: Sopel 8 upgrade guide

The result of lengthy collaboration. Commit metadata does not *quite*
reflect the actual authorship timeline... ;-)

Co-authored-by: James <snoopjedi@gmail.com>
Co-authored-by: mal <mal@sec.gd>

Exirel commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### Enums everywhere
+
+_(Insert that classic Buzz Lightyear meme here, if you want.)_
```
Would be funny to keep it at a text line, and nothing else.

dgw replied:
That's exactly what this text line is supposed to be, actually 😁🤣

Exirel reviewed (commented):
Looks good to me.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+been removed. In exchange, known issues sometimes preventing use of Sopel 7 with
+Python 3.11+ have been fixed.
+
+During the lifecycle of Sopel 8.x, we will add new Python releases to our test
+suite as soon as possible. We will keep testing against end-of-life Python
```
> as soon as possible

Maaay be a little too strong of a promise, could soften it to say that we'll make a best-effort attempt to track the leading edge (more specifically, beta/RC releases of CPython)

I said a bunch about this in my initial review which I think boils down to: _"Sopel should have a declared policy about supported versions"_

dgw replied:
According to my review UI, this is the last comment left (for now). I was going to make this suggestion in a new thread across the whole paragraph's worth of lines, but the wording I ended up with happened to fit the changes into just these two that were already selected (!)

```suggestion
During the lifecycle of Sopel 8.x, we will do our best to test against new
Python releases as they come. We will keep testing against end-of-life Python
```

half-duplex replied:
Seconding best-effort, unless someone's writing checks.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+Continuing our push to clarify the difference between a "module" (which is a
+Python concept) and a "plugin" (which is a special kind of "module" that can
+extend Sopel with new functionality), Sopel 8 no longer loads plugins from the
+`<homedir>/modules` directory by default.
```
I inverted the order of the clauses here and added a migration instruction. The migration guide reader probably mostly cares about the change and not the rationale behind the change, although it's good to have both.

Also added a link to the official definition of a module for the curious reader, but that's real nit-picking stuff.

```suggestion
Sopel 8 no longer loads plugins from the `<homedir>/modules` directory by default.
To migrate, rename this directory to `plugins/` or declare it in the `extra` field of the
core configuration section.

This is part of our efforts to clarify the difference between a "module" (which is a
[Python concept][py-module]) and a "plugin" (which is a special kind of "module" that
can extend Sopel with new functionality).

  [py-module]: https://docs.python.org/3/glossary.html#term-module
```

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### Command names are now literal
+
+To prevent conflicts with the default capture groups Sopel defines for commands
+and prepare for future enhancements to plugin/command management, characters
+with special meaning in regular expressions are now escaped in command names.
```
Hmm, this feels like it could be misread as describing a smaller feature area than it is. In other words, this is saying that certain parts of regex patterns no longer work, but the heft of the change is that the `@command` input itself is _no longer a pattern_.

Maybe it's simpler to say that users can't pass a pattern of any kind anymore?

dgw replied:
I hoped the section title would carry that, but you might have a point. Worked up an alternate wording:

```suggestion
To prevent conflicts with the default capture groups Sopel defines for commands
and prepare for future enhancements to plugin/command management, regex command
names are no longer supported. Characters in command names that have special
meaning for regular expressions will be escaped.
```

SnoopJ replied:
That does still read to me as being in rationale/change order, but the clarification is a good one :+1:

dgw replied:
Trying again:

```suggestion
Sopel 8 no longer supports regular expressions in command names, which are now
[escaped][re.escape] when a plugin is loaded. This prevents interference with
the default capture groups for commands, and also clears the way for
enhancements to plugin/command management.
```

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+### IRC event changes
+
+The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
+clarifications from the living-specification project at modern.ircdocs.horse.
```
Might as well link to the relevant spec here (assuming there's a suitable anchor)

dgw replied:
This will be a fun one to do in suggestions because it needs changes split across locations, but I'll give it a shot.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+# Sopel 8.0 Migration
+
+Version 8.0 brings Sopel into the modern era of Python, finally dropping support
+for Python 2 and all end-of-life Python 3 releases as of December 2023. This let
+us make significant updates under the hood to future-proof things like the IRC
+connection backend and event system.
```
From my initial review: would prefer to reword this, and add a link to the official CPython docs with a table/Gantt chart of which versions are in which part of the lifecycle.

```suggestion
Version 8.0 brings Sopel into the modern era of Python, finally dropping support
for [end-of-life](python-devguide-versions) Python releases as of December 2023,
including Python 2 and several older versions of Python 3. This allows us to make
significant improvements under the hood that were held back by these legacy
versions, e.g. the IRC connection backend and Sopel's event system.

  [python-devguide-versions]: https://devguide.python.org/versions/
```

dgw replied:
Will accept with a slight edit: indenting the link definition to match others in the file.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+#### `reload` plugin
+
+The `.update` command has been removed from the `reload` plugin. Its functioning
+relied on running Sopel in a way that isn't officially supported, so we chose to
+remove it.
```
We should probably point users who used to depend on `.update` towards `.reload` or restarting Sopel. Guessing that the latter is the preferred option.

dgw replied:
This is a good idea; I ran with it.

I'm reasonably sure about the URLs, but not 100% certain as the Netlify configuration could differ slightly between the `unstable.docs` for 8.0.0 and the main site.

```suggestion
The `.update` command has been removed from the `reload` plugin. It relied on
running Sopel in a way that we don't officially support, so we chose to remove
the command.

To replace `.update`, if you were using it, we recommend writing an external
script that [stops Sopel][cli-stop], updates the components you want, and then
[starts Sopel][cli-start] again.

  [cli-stop]: https://sopel.chat/docs/run/cli.html#sopel-stop
  [cli-start]: https://sopel.chat/docs/run/cli.html#sopel-start
```

dgw replied:
(We can avoid the uncertainty about how the final URLs will look by skipping the links for now, mentioning `sopel start` and `sopel stop` commands directly, and linkifying them post-release if desired.)

Exirel replied:
The `.reload`  should never ever be used or promoted. We collectively spent too much time on trying to figure out a way to make it work and it still doesn't.

dgw replied:
@Exirel Are you saying not to mention the removal of `.update` at all? I'm not sure what change (if any) you are indirectly suggesting to the proposed migration guide.

Exirel replied:
It's a reply to SnoopJ's initial comment to talk about `.reload`.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+This is a good time to remind you that your plugins can specify both minimum and
+maximum Sopel versions in their requirements. If all plugins a bot owner wishes
+to install contain this metadata, they can ask `pip` to install Sopel and all of
+its plugins in a single command, and `pip`'s dependency solver will figure out
+which version of Sopel satisfies all of the plugins' needs.
```
We should probably clarify that plugins installed by themselves can therefore _change_ the version of Sopel that is installed, and users who do _not_ want to let their Sopel version change should include `sopel` in every install command.

The statement about `pip` figuring out the needed version of Sopel is a liiitle misleading here, since what `pip` does is solve _the entire graph_, which could result in using an older version of a plugin in cases where that's needed. I'm nit-picking, though.

dgw replied:
You are nitpicking, and I'm going to avoid spending time on it unless we specifically [decide](https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548727505) to keep this section. 😁

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+its plugins in a single command, and `pip`'s dependency solver will figure out
+which version of Sopel satisfies all of the plugins' needs.
+
+We do everything we can to keep breaking API changes to *only* major versions,
+so a version range like `sopel>=7.1,<9` is typically safe—at least the earliest
```
`>=7.1` is kinda surprising to me here, why not `sopel>=8,<9` and clarify that plugin devs who want more nuanced control over the requirements should figure out what works for them?

dgw replied:
I'm going to suggest against an expanded range of lines, since I can't actually implement your clarification idea against only these two. 😸

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+its plugins in a single command, and `pip`'s dependency solver will figure out
+which version of Sopel satisfies all of the plugins' needs.
+
+We do everything we can to keep breaking API changes to *only* major versions,
+so a version range like `sopel>=7.1,<9` is typically safe—at least the earliest
+Sopel version your plugin supports, up to (but excluding) the next unreleased
+major version.
+
+In the vast majority of cases, removed API features will first go through a
+deprecation period, during which Sopel will log warnings whenever the deprecated
+functionality is used by a plugin. Rarely, we might need to remove an API
+feature without going through this process—but that's a last-resort option.
```
Is it perhaps worth mentioning the `Breaking Change` label on GitHub as the typical marker for breakage?

dgw replied:
I'd hope anyone who follows the issue tracker has already figured that one out. But if we _don't_ delete this section I'll give this some more thought as we edit it.

dgw replied:
Removed the whole block about plugin versioning, mooting this conversation 😺

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+We do everything we can to keep breaking API changes to *only* major versions,
+so a version range like `sopel>=7.1,<9` is typically safe—at least the earliest
+Sopel version your plugin supports, up to (but excluding) the next unreleased
+major version.
+
+In the vast majority of cases, removed API features will first go through a
+deprecation period, during which Sopel will log warnings whenever the deprecated
+functionality is used by a plugin. Rarely, we might need to remove an API
+feature without going through this process—but that's a last-resort option.
```
Overall, this section feels more like general guidance on plugin authorship than it does a guide to migrating between <8 and 8.0.

dgw replied:
Before I get to the other comments in this section (the lede of "Sopel 8 API changes"), it might indeed make more sense to just drop it and go directly from the heading to the first subsection:

```md
## Sopel 8 API changes

### Command names are now literal
```

Exirel replied:
I think I agree yes. Do we need a way to talk about plugin management in some way or another? Yes. Is it the right place to do so? Maybe not. A lot of good ideas in here in any case.

dgw replied:
I removed this section from the newest upgrade guide revision, because I agree this isn't the place to talk about plugin versioning & dependency management. sopel-irc/sopel#2604 will carry the torch from here.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+`trigger.time` is now an "aware" `datetime` object, meaning it has a UTC offset
+associated with it. Comparison or arithmetic operations between "aware" and
+"naive" `datetime` objects are not allowed; code that manipulates `trigger.time`
+values will need to be updated for Sopel 8.
```
We should probably expand a little bit on _how_ it should be updated or at least point the reader in the right direction if they haven't met timezone-aware objects before.

dgw replied:
Going to get into this in a follow-up to your other comment on this block, which selects more lines and lends itself better to writing suggestions.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### Time-handling changes
+
+`trigger.time` is now an "aware" `datetime` object, meaning it has a UTC offset
+associated with it. Comparison or arithmetic operations between "aware" and
+"naive" `datetime` objects are not allowed; code that manipulates `trigger.time`
+values will need to be updated for Sopel 8.
```
Good spot for an official docs link

```suggestion
`trigger.time` is now an ["aware" `datetime` object][py-aware], meaning it has a
UTC offset associated with it. Comparison or arithmetic operations between
"aware" and "naive" `datetime` objects are not allowed; code that manipulates
`trigger.time` values will need to be updated for Sopel 8.

  [py-aware]: https://docs.python.org/3.11/library/datetime.html#aware-and-naive-objects
```

dgw replied:
I edited the above to be syntactically correct Markdown, and fix the text wrapping, but let me go further based on feedback from your other line note (https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548728017):

```suggestion
`trigger.time` is now an ["aware" `datetime` object][py-aware], meaning it has a
UTC offset associated with it. Comparison or arithmetic operations between
"aware" and "naive" `datetime` objects are not allowed; code that manipulates
`trigger.time` values will need to make sure "naive" `datetime` objects created,
or retrieved from sources, outside the bot are converted to "aware" values.

  [py-aware]: https://docs.python.org/3.11/library/datetime.html#aware-and-naive-objects
```

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+"naive" `datetime` objects are not allowed; code that manipulates `trigger.time`
+values will need to be updated for Sopel 8.
+
+The fallback format string used by `sopel.tools.time.format_time()` if no other
+source provides one changed from outputting the timezone name (`UTC`) to the UTC
+offset (`+0000`).
```
Nit-pick: this sentence is a little hard to read even as a native speaker, feels like it could use a rewrite.

dgw replied:
I can see it… Woof, past-me. Retry, longer but hopefully clearer:

```suggestion
The fallback format string used by `sopel.tools.time.format_time()` changed from
outputting the timezone name (`UTC`) to the UTC offset (`+0000`). This format
string is used only if there is no preferred format set for the nick, channel,
_and/or_ config passed to the function.
```

SnoopJ replied:
Much more legible!

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+source provides one changed from outputting the timezone name (`UTC`) to the UTC
+offset (`+0000`).
+
+`format_time()`'s handling of "aware" and "naive" `datetime`s was also improved,
+but those changes should be transparent to both users and plugin developers.
```
If they're transparent, do we even want to mention them in the migration guide?

dgw replied:
We want to know about any cases plugin developers find that turn out not to have been transparent. Maybe that should be more explicit?

```suggestion
`format_time()`'s handling of "aware" and "naive" `datetime`s was also improved,
but those changes should be transparent to both users and plugin developers. Let
us know if you encounter situations where `format_time()` behaves surprisingly.
```

Exirel replied:
Yeah I think it's good to talk about it, so developers are more cautious with their usage of time-related functions.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### IRC event changes
+
+The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
```
Updated how? IIRC from my initial review, these names changed and we should probably state the old names here.

dgw replied:
The `invitelist-madness` link definition below covers it. If you have a better place to link it, closer to the top of this section, I'm all eyes/ears/whatever metaphor you want to use in this situation. 😁

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### IRC connection status monitoring
+
+Previously a simple `bool` value, `bot.connection_registered` is now a property
+combining status information from across the bot's subsystems to answer the very
+important question, "Is the `bot` actually registered to an IRC network?" That
+is, Sopel not only has a _connection_ open, but the IRC server has _accepted_
+the bot as a _client_ and it's possible to _send commands_ to the IRC server.
```
Sharing here a reword I did of this in my initial review some months ago. I don't feel as strongly now as I did about it then, though.

```suggestion
`bot.connection_registered` now indicates that a _connection_ is open, the 
IRC server has _accepted_ the bot as a client, and commands can be sent.
Previously, this boolean value indicated only that a connection was open. This
attribute is now read-only and cannot be modified.
```

dgw replied:
Going to take my own stab at this one, using yours as a starting point:

```suggestion
`bot.connection_registered` now indicates that a _connection_ is open, the 
IRC server has _accepted_ the bot as a client, and the bot can _send commands_.
Previously, this boolean value indicated only that a connection was open.
```

Removed the disclaimer that it's read-only _now_ because no one outside core should have been writing to it anyway. We've just codified the implicit contract by making it into a `@property`.

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+#### Only set the `sender` when it makes sense
+
+For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
+a channel or other clearly defined _context_, the `trigger.sender` value that
+plugin callables saw in Sopel 7 was unpredictable at best.
+
+Sopel 8.0 now _only_ sets the `trigger.sender` if the triggering event warrants
+it. In all other cases, the `sender` will be `None` and plugin code should use
+the `trigger.args` list to retrieve all information about the event.
```
Sharing a rewrite of this section I did in my initial review. Main change is that it now lists out the events that do set `trigger.sender`

```suggestion
Sopel 8.0 now _only_ sets `trigger.sender` for the following events:
  * `INVITE`
  * `JOIN`
  * `KICK`
  * `MODE`
  * `NOTICE`
  * `PART`
  * `PRIVMSG`
  * `TOPIC`

In all other cases, the `sender` will be `None` and plugin code should use the
`trigger.args` list to retrieve all information about the event.

For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
a channel or other clearly defined _context_, the `trigger.sender` value that
plugin callables saw in Sopel 7 was unpredictable at best.
```

dgw replied:
Once again, using your suggestion as a starting point for further polishing:

```suggestion
Sopel 8.0 now _only_ sets `trigger.sender` for the following events:

  * `INVITE`
  * `JOIN`
  * `KICK`
  * `MODE`
  * `NOTICE`
  * `PART`
  * `PRIVMSG`
  * `TOPIC`

In all other cases, the `sender` will be `None` and plugin code should use the
`trigger.args` list to retrieve all information about the event.

For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
a channel or other clearly defined _context_, the `trigger.sender` value that
plugin callables saw in Sopel 7 was unpredictable at best. This change is
intended to make it easier to write correct behavior going forward.
```

SnoopJ commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+### IRC privilege requirement changes
+
+In Sopel 8.0, the `@sopel.plugin.require_privilege()` decorator now implies
+`@sopel.plugin.require_chanmsg()`, and the decorated callable will not run if
+triggered outside of a channel. Previously, `require_privilege` restrictions
+were simply ignored, and the callable would run anyway.
```
Not really feedback on the migration guide, but when reading this originally, it struck me that I wished these decorators referred to "channel privilege" to better align with "bot privilege". Maybe something to consider for 9.0

dgw replied:
If this is something that hasn't been brought up yet (or at least recently), a better place to explore this would be the main issue tracker (see past issues with "RFC" in the title), or even better to start off in the IRC channel and flesh out the idea a little before writing it up as a planning issue.

`require_privilege` is [9 years old](https://github.com/sopel-irc/sopel/commit/3feb82ab24891fd5a4a6449f3e7e64cba4fd4865) and I wouldn't want to move too quickly toward removing it, but it could be converted to a compatibility alias pretty easily. (That's a structural topic for another day, too: Possibly making `sopel.plugin` into a full-fat submodule that lives in `sopel/plugin/__init__.py` so we can define compatibility shims in someplace like `sopel.plugin.compat`.)

SnoopJ replied:
Now tracking as: https://github.com/sopel-irc/sopel/issues/2605

SnoopJ reviewed (approved):
Sharing some long-overdue feedback, most of it authored in December for an earlier draft that was similar to this.

Even though I have some requested changes, I'm approving this as-is, because clearly none of my feedback is so important that it couldn't wait several months. At this point, our best feedback will probably come from users, and we just want to get 8.0 out the door.

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+its plugins in a single command, and `pip`'s dependency solver will figure out
+which version of Sopel satisfies all of the plugins' needs.
+
+We do everything we can to keep breaking API changes to *only* major versions,
+so a version range like `sopel>=7.1,<9` is typically safe—at least the earliest
+Sopel version your plugin supports, up to (but excluding) the next unreleased
+major version.
```
This goes with https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548726609 — I needed the extra lines to make the suggestion work.

```suggestion
We do everything we can to keep breaking API changes to *only* major versions,
so a version range like `sopel>=8.0,<9` is typically safe. If you've tested
against older versions of Sopel, you can mark it as compatible with the earliest
tested version up to (but excluding) the next unreleased major version.
```

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+`sopel.tools.time.validate_timezone()` now also raises a `ValueError` even if
+the timezone to be validated is `None`. This was previously a special case that
+_returned_ `None`.
+
+### Database changes
+
+`bot.db.get_nick_id()` no longer creates a new ID by default if its `create`
+parameter is unspecified.
+
+### IRC event changes
+
+The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
+clarifications from the living-specification project at modern.ircdocs.horse.
+`RPL_INVEXLIST` and `RPL_ENDOFINVEXLIST` were added to the list of named events
+that Sopel knows about.
+
+Plugin code might need to be updated as a result of these changes, but keep in
+mind that [different IRC servers mean different things][invitelist-madness] when
+they send these events. If in doubt, specify a raw numeric value (e.g. `'346'`).
+Writing truly cross-network plugins that react to these events should be
+undertaken _very_ carefully.
+
+  [invitelist-madness]: https://github.com/ircdocs/modern-irc/issues/42
```
This addresses multiple related comments, https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548729687 and https://github.com/sopel-irc/sopel.chat/pull/46/files#r1548715208

```suggestion
The event names `RPL_INVITELIST` and `RPL_ENDOFINVITELIST` have been updated per
[clarifications][invitelist-madness] from the "Modern IRC" living-specification project.
`RPL_INVEXLIST` and `RPL_ENDOFINVEXLIST` were added to the list of named events
that Sopel knows about.

Plugin code might need to be updated as a result of these changes, but keep in
mind that [different IRC servers mean different things][invitelist-warning] when
they send these events. If in doubt, specify a raw numeric value (e.g. `'346'`).
Writing truly cross-network plugins that react to these events should be
undertaken _very_ carefully.

  [invitelist-madness]: https://github.com/ircdocs/modern-irc/issues/42
  [invitelist-warning]: https://modern.ircdocs.horse/#invite-list
```

SnoopJ replied:
Fantastic :smile_cat:

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+received [via a socket][sopel-sockmsg]. `if not bot.connection_registered`, the
+output can be skipped, retried after a delay, etc.
+
+_Hopefully no one was overwriting the old simple attribute in plugin code, but
+now they **can't** do that._
+
```
To go along with [my other suggestion](https://github.com/sopel-irc/sopel.chat/pull/46/files#r1564816742), downplaying the fact that plugins used to have the option to be naughty:

```suggestion
```

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,428 @@
+
+#### What's a "module", doc?
+
+Sopel 8 no longer loads plugins from the `<homedir>/modules` directory by default.
+To migrate, rename this directory to `plugins/` or declare it in the `extra` field of the
+core configuration section.
+
+This is part of our efforts to clarify the difference between a "module" (which is a
+[Python concept][py-module]) and a "plugin" (which is a special kind of "module" that
+can extend Sopel with new functionality).
+
+  [py-module]: https://docs.python.org/3/glossary.html#term-module
```
The new `core-section-plugins` link doesn't exist yet on the live site, but that page was moved for 8.0 in sopel-irc/sopel#2533.

```suggestion
Sopel 8 no longer loads plugins from the `<homedir>/modules` directory by default.
To migrate, rename this directory to `plugins/` or declare it in the `extra` field of
[the `[core]` configuration section][core-section-plugins].

This is part of our efforts to clarify the difference between a "module" (which is a
[Python concept][py-module]) and a "plugin" (which is a special kind of "module" that
can extend Sopel with new functionality).

  [core-section-plugins]: https://sopel.chat/docs/run/configuration.html#plugins
  [py-module]: https://docs.python.org/3/glossary.html#term-module
```

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,428 @@
+and prepare for future enhancements to plugin/command management, characters
+with special meaning in regular expressions are now escaped in command names.
+
+Documentation for `@sopel.plugin.command()` suggests alternatives based on the
+use case you need to adapt for Sopel 8.
```
```suggestion
Documentation for `@sopel.plugin.command()` suggests alternatives based on the
use case you need to adapt for Sopel 8.

  [re.escape][https://docs.python.org/3.11/library/re.html#re.escape]
```

dgw commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,420 @@
+
+During the lifecycle of Sopel 8.x, we will add new Python releases to our test
+suite as soon as possible. We will keep testing against end-of-life Python
+versions unless and until it becomes technically impossible to do so. However,
+if testing against an EOL Python version does become impossible, we will drop it
+in the next maintenance release of Sopel.
```
No one gave me feedback on this part; I just thought it could be worded better while looking at the other comments around it.

```suggestion
versions unless and until it becomes technically impossible to do so—in which
case Sopel will officially drop support for that EOL Python release in its next
maintenance release.
```

half-duplex replied:
Personally, "technically impossible" is strong. "impractical"?

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+freed up for use by bespoke plugins. `wikipedia` functionality is now invoked by
+the `.wikipedia` command, or `.wp` for short.
+
+The deprecated `lang_per_channel` config-file setting has been removed.
```
Mention `.wpclang ja`, or link .wpclang doc?

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+
+The following parts of Sopel's API are now expressed as some type of `Enum`:
+
+* `sopel.formatting.colors`
+* `sopel.tools.events`
```
If the names didn't change and this doesn't require a change in normal plugin code, say so (or leave this off the upgrade guide?), otherwise link docs?

dgw replied:
I spent probably half an hour on this section, with the rewrite to come in my next push.

Definitely keeping all commits separate for now so it's easier for you and the other reviewers to see what changed!

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+
+### Time-handling changes
+
+`trigger.time` is now an ["aware" `datetime` object][py-aware], meaning it has a
+UTC offset associated with it. Comparison or arithmetic operations between
+"aware" and "naive" `datetime` objects are not allowed; code that manipulates
+`trigger.time` values will need to make sure "naive" `datetime` objects created,
+or retrieved from sources, outside the bot are converted to "aware" values.
+
+  [py-aware]: https://docs.python.org/3.11/library/datetime.html#aware-and-naive-objects
```
Since this one had no deprecation period and is somewhat likely to require code changes by anything that uses it, consider calling attention to this section (whether a :warning: block or a note in a ToC or tl;dr)

dgw replied:
Yeah, this one's tricky. We have very limited tools in the Jekyll/Markdown sections of the website as compared to Sphinx's robust set of admonition types (warning boxes etc.).

I will move it up to be the second subhead in this section (command names no longer accepting regex is also breaking) and see what I can find to call it out more strongly within the limitations of Jekyll.

Consider this further motivation for us to move the rest of the site into Sphinx, as we've already started doing…

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+`bot.say()`, `bot.reply()`, etc. in direct response to a `trigger` and without
+overriding the default `destination` will most likely get the expected behavior.
+
+[`sopel-remind`][sopel-remind] is an example of a plugin that this change might
+surprise, since the naïve implementation of such a reminder plugin would simply
+store `(trigger.nick, trigger.sender, parsed_message)` and send
+`bot.reply(parsed_message, trigger.sender, trigger.nick)` after the specified
+amount of time has passed. The `trigger.status_prefix` is lost, and the reminder
+sent to the entire channel, rather than the status-limited subset of users who
+could have seen the original command.
```
Clarify that sopel-remind does this *correctly*, at least `s/a plugin/the kind of plugin/`.

dgw replied:
It _doesn't_ handle this correctly, yet. sopel-irc/sopel-remind#10

half-duplex replied:
Oh, fun.

half-duplex replied:
Oh, fun.

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+  `bot.channels`)
+* `bot.msg()` method (use `bot.say()`)
+* `bot.register()` & `bot.unregister()` methods
+  * Pretty much no one should have needed to use these in a plugin. If for some
+    reason you did, you'll be able to find the relevant documentation chapter
+    that describes their replacements yourself.
```
"Plugins don't normally [manually whatever], but if yours did, see [link]."
More concise and not sassy

dgw replied:
I'm fine with reducing the sass and will rewrite this, but let's not link the replacement in classes from `sopel.plugins.handlers`; that's an internal area of the API.

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+  * As a reminder, plugins **should** no longer use the above-mentioned methods;
+    the `bot.rules.check_url_callbacks(bot, url)` method now works very well
+    with the `@sopel.plugin.url()` decorator
+* `bot.privileges` channel information (deprecated since Sopel 6.2; replaced by
+  `bot.channels`)
+* `bot.msg()` method (use `bot.say()`)
```
Link docs, for these and other "(use x)" below. If msg/say syntax is the same, say "renamed" to save readers having to go check their arguments.

dgw replied:
> If msg/say syntax is the same

It's not.

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -0,0 +1,435 @@
+
+Previously deprecated parts of Sopel's API have been removed, including:
+
+* `bot.memory['url_callbacks']` is no longer created or populated by the bot
+  * Tracking this information moved to an internal property, in preparation for
+    the removal of `bot.register_url_callback()`, `bot.search_url_callbacks()`,
+    and `bot.unregister_url_callback()` in Sopel 9.0
+  * As a reminder, plugins **should** no longer use the above-mentioned methods;
+    the `bot.rules.check_url_callbacks(bot, url)` method now works very well
+    with the `@sopel.plugin.url()` decorator
```
```md
* `bot.memory['url_callbacks']` is no longer created or populated by the bot
  * `bot.register_url_callback()` and `bot.unregister_url_callback()` are deprecated.
  * Use [`@sopel.plugin.url()`][url-decorator] or [`@sopel.plugin.url_lazy()`][url-lazy-decorator] instead.
  * `bot.search_url_callbacks()` is deprecated, replace with [`bot.rules.check_url_callbacks(bot, url)`][check-url-callbacks]
    
  [url-decorator]: https://docs.sopel.chat/plugin/decorators.html#sopel.plugin.url
  [url-lazy-decorator]: https://docs.sopel.chat/plugin/decorators.html#sopel.plugin.url_lazy
  [check-url-callbacks]: https://docs.sopel.chat/package/plugins/rules.html#sopel.plugins.rules.Manager.check_url_callback
```

The last doc link is... not super helpful for understanding you need to do `bot.rules...`

dgw replied:
I adapted much of this. Couldn't apply directly anyway, for multiple reasons.

dgw commented:
@half-duplex Bunch of changes pushed up in response to your feedback 😸

half-duplex commented on _upgrading/sopel-8.md:
```diff
@@ -182,18 +186,49 @@ use case you need to adapt for Sopel 8.
 * `sopel.formatting.colors`
+  * Class members' names and values remain unchanged
+* `sopel.plugin.OP`, `sopel.plugin.VOICE`, etc.
+  * These constants became shortcuts to equivalent members of a new
+    `sopel.privileges.AccessLevel` enumeration, with the same values
```
If the original names are deprecated, "These deprecated constants are now..."?

dgw replied:
Not deprecated _yet_ 😁

half-duplex replied:
...oh

dgw commented:
Seeing no further objections (feedback), I have started getting this ready to merge.

Will most likely create a new PR to hold all of the 8.0-related updates and test integrating all the different parts (upgrade guide, updated submodule docs, updated redirects, etc.) before final merge.

dgw commented:
Current status: Trying to convince Netlify that it should actually _show_ the deploy log so I have a chance to fix whatever went wrong after pulling in sopel-irc/sopel#2606 to test the new docs. It gets stuck like this and never loads the log:

![image](https://github.com/sopel-irc/sopel.chat/assets/164140/424ae81e-9b84-4e6a-b229-007f0351b4b7)

----

**Edit:** On a fresh branch that only touches `_sopel/`, it still fails. Was able to get the logs in a different browser but they are not helpful. It doesn't show any of the build script's output, only a non-zero exit code:

```
7:05:13 PM: Failed during stage 'building site': Build script returned non-zero exit code: 2 (https://ntl.fyi/exit-code-2)
```

![image](https://github.com/sopel-irc/sopel.chat/assets/164140/0db1e1cc-8ddb-452f-b8d4-5993c912b277)

**I will revert this branch to the known-working upgrade guide and continue debugging `_sopel/` in a new PR.**

dgw commented:
Base branch changed to #47. I will most likely revert the test-merge of #38 there and perform a manual three-way merge of both PRs into the integration branch/PR—hoping that it goes well.
